queue/driver_mongo.go

```diff
@@ -880,14 +799,115 @@ RETRY:
 			}
 
 			if job != nil {
-				break RETRY
+				return job
+			}
+
+			timer.Reset(time.Duration(misses * rand.Intn(int(d.opts.WaitInterval))))
+		}
+	}
+}
+
+// dispatchAttemptInfo contains aggregate statistics on an attempt to dispatch
+// jobs.
+type dispatchAttemptInfo struct {
+	// skips is the number of times it encountered a job that was not yet ready
+	// to dispatch.
+	skips int
+	// misses is the number of times the dispatcher attempted to dispatch a job
+	// but failed.
+	misses int
+}
+
+// tryDispatchJob takes an iterator over the candidate Amboy jobs and attempts
+// to dispatch one of them. If it succeeds, it returns the successfully
+// dispatched job.
+func (d *mongoDriver) tryDispatchJob(ctx context.Context, iter *mongo.Cursor, startAt time.Time) (amboy.Job, dispatchAttemptInfo) {
+	var dispatchInfo dispatchAttemptInfo
+	for iter.Next(ctx) {
+		j := &registry.JobInterchange{}
+		if err := iter.Decode(j); err != nil {
+			grip.Warning(message.WrapError(err, message.Fields{
+				"id":            d.instanceID,
+				"service":       "amboy.queue.mdb",
+				"operation":     "converting next job",
+				"message":       "problem reading document from cursor",
+				"is_group":      d.opts.UseGroups,
+				"group":         d.opts.GroupName,
+				"duration_secs": time.Since(startAt).Seconds(),
+			}))
+			// try for the next thing in the iterator if we can
+			continue
+		}
+
+		job, err := j.Resolve(d.opts.Format)
+		if err != nil {
+			grip.Warning(message.WrapError(err, message.Fields{
+				"id":            d.instanceID,
+				"service":       "amboy.queue.mdb",
+				"operation":     "converting document",
+				"message":       "problem converting job from intermediate form",
+				"is_group":      d.opts.UseGroups,
+				"group":         d.opts.GroupName,
+				"duration_secs": time.Since(startAt).Seconds(),
+			}))
+			// try for the next thing in the iterator if we can
```

Comment:
I think you need to set `job = nil` here and before the rest of the `continue` statements.

Reply:
We don't because the job is not declared outside the loop.  The only way it could return the job is on L907, after the job is already dispatched.

Reply:
Ah, right.